### DISCUSSION TITLE
In what scenarios where custom server is better to Kestrel?

### DISCUSSION THREAD

**Discussion by ladeak** [dotnet/aspnetcore#58952] [General] [answered]
In what scenarios where custom server is better to Kestrel?
- I am wondering about niche protocol such gemini
- or specific use-cases with existing protocols (ie. performance/stability etc.)
- or specific connection types (socket, file sockets/pipes, etc.)?

I wonder if @Tratcher has some ideas.

**Comment by gfoidl** [comment, marked as answer]
In some 1:1 communication with e.g. micro-controllers Kestrel may be overkill. There a custom TCP-server can be a better fit.

When it's 1:n (1 server, n clients), as well as public facing a full-flegded and proven server like Kestrel is -- of course -- more suitable.
